ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs

```diff
@@ -0,0 +1,698 @@
+  -> Command Concrete
+  -> m (Response Concrete)
+semantics vars = \case
+    Disconnect peer -> do
+        atomically $ do
+            modifyTVar varCandidates $ Map.delete peer
+            modifyTVar varIdlers     $ Set.delete peer
+        pure Unit
+    ExtendSelection sdel -> do
+        atomically $ do
+            Selection b s <- readTVar varSelection
+            writeTVar varSelection $! Selection (b + 1) (s + sdel)
+        pure Unit
+    ModifyCandidate peer bdel -> do
+        atomically $ do
+
+            modifyTVar varIdlers $ Set.delete peer
+
+            v <- (Map.! peer) <$> readTVar varCandidates
+            Candidate b <- readTVar v
+            writeTVar v $! Candidate (b + bdel)
+
+        pure Unit
+    NewCandidate peer bdel -> do
+        atomically $ do
+            Selection b _s <- readTVar varSelection
+            v <- newTVar $! Candidate (b + bdel)
+            modifyTVar varCandidates $ Map.insert peer v
+        pure Unit
+    ReadJudgment -> do
+        yield10
+        fmap ReadThisJudgment $ atomically $ readTVar varJudgment
+    ReadMarker -> do
+        yield10
+        fmap ReadThisMarker $ atomically $ readTVar varMarker
+    StartIdling peer -> do
+        atomically $ modifyTVar varIdlers $ Set.insert peer
+        pure Unit
+    TimePasses dur -> do
+        SI.threadDelay (0.1 * fromIntegral dur)
+        pure Unit
+  where
+    Vars varSelection varCandidates varIdlers varJudgment varMarker = vars
+
+type Model :: (Type -> Type) -> Type
+data Model r = Model {
+    mCandidates :: Map.Map UpstreamPeer Candidate
+  ,
+    mClock :: SI.Time
+  ,
+    mIdlers :: Set.Set UpstreamPeer
+  ,
+    mPrev :: WhetherPrevTimePasses
+  ,
+    mSelection :: Selection
+  ,
+    mState :: ModelState
+  }
+  deriving (Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+initModel :: LedgerStateJudgement -> Model r
+initModel j = Model {
+    mCandidates = Map.empty
+  ,
+    mClock = SI.Time 0
+  ,
+    mIdlers = Set.empty
+  ,
+    mPrev = WhetherPrevTimePasses True
+  ,
+    mSelection = Selection 0 s
+  ,
+    mState = case j of
+        TooOld      -> ModelTooOld
+        YoungEnough -> ModelYoungEnough (SI.Time (-10000))
+  }
+  where
+    s = S $ case j of
+        TooOld      -> (-11)
+        YoungEnough -> 0
+
+precondition :: Model Symbolic -> Command Symbolic -> QSM.Logic
+precondition model = \case
+    ExtendSelection _sdel -> QSM.Boolean $
+        selectionIsBehind model
+    Disconnect peer -> QSM.Boolean $
+        peer `Map.member` cands
+    ModifyCandidate peer _bdel -> QSM.Boolean $
+        peer `Map.member` cands
+    NewCandidate peer _bdel -> QSM.Boolean $
+        (not $ peer `Map.member` cands)
+    ReadJudgment ->
+        QSM.Top
+    ReadMarker ->
+        QSM.Top
+    StartIdling peer -> QSM.Boolean $
+        (peer `Map.member` cands)
+     &&
+        (peer `Set.notMember` idlers)
+    TimePasses dur -> QSM.Boolean $
+        (0 < dur)
+     &&
+        (boringDur model dur)
```

Comment:
I was avoiding that, since the annotations are unused by this test.

However, I realized that they _are_ used when you start manually manipulating repros. So now I've embraced the mini DSL.